### PR TITLE
TM: Py, SDR, UnitTests

### LINKED ISSUE #243: Connections - Python Bindings
Python should be able to use the C++ Connections class.  This is important for experimentation, to combine the ease-of-use of python with the speed of C++.

Results:
* Python user can create, setup, and run the connections in a simple example.
* Python user can manipulate synapses, segments, cells & all of their data fields.
* Python connections module should be easy & intuitive to use.

### PR DESCRIPTION
**Pull request by ctrl-z-9000-times**
For issue #289 
Fixes #243 

- [x] Connections class python bindings
- [x] Unit tests for Connections from python
- [x] TM python bindings
- [x] TM unit tests from python

### CODE REVIEW

**Comment by breznak** [comment]
My problem here is with layering, 

- I'm :+1: for the TM tests and bindings. 
- :-1: for exposing Connections. As I've stated elsewhere, my worry is any standartization (ie via bindings) of the Connections API. That is exactly the same situation as with SparseMatrix; COnnections is used by public facing APIs of SP, TM, CP,... not giving access to it(Conn) to users would give the developers (of nupic.cpp) free hands at changing, or at some point ditching, that framework. 
  - an argument for exposing connections would be allowing new developement
    - counter, for research prototype, wouldn't Connections in pure python for py dev, and c++ Conn for c++ be enough? 
    - if we agree we still want bindings for connections, it should be marked with a warning that the API may break (for both py, c++)

File: bindings/py/tests/algorithms/temporal_memory_test.py [hunk 1]
```diff
@@ -0,0 +1,1255 @@
...
+      proto2 = TemporalMemoryProto_capnp.TemporalMemoryProto.read(f)
+
+    # Load the deserialized proto
+    tm2 = TemporalMemory.read(proto2)
+
+    self.assertEqual(tm1, tm2)
+    self.serializationTestVerify(tm2)
+
+
+  @unittest.skip("Manually enable this when you want to use it.")
```
Comment (breznak) [inline bindings/py/tests/algorithms/temporal_memory_test.py]: remove these 2 tests, they use capnp anyway

File: src/nupic/algorithms/TemporalMemory.hpp [hunk 2]
```diff
@@ -512,6 +512,7 @@ using nupic::algorithms::connections::Synapse;
 
 public:
   Connections connections;
+  const UInt &extra = extra_;
```
Comment (breznak) [inline src/nupic/algorithms/TemporalMemory.hpp]: TODO document

**Comment by ctrl-z-9000-times** [comment]
Sooner or later, `py_Connections` is going to happen.

The connections class is literally the most important piece of code in this repo, everything else is secondary.  The _only_ advantage of C++ over python is speed and an HTM spends ~90% of its time in the connections class.

File: bindings/py/cpp_src/bindings/algorithms/py_HTM.cpp [hunk 3]
```diff
@@ -181,6 +188,15 @@ using nupic::algorithms::connections::Permanence;
             self.compute(activeColumns.size(), get_it(activeColumns), learn);
         }, "", py::arg("activeColumns"), py::arg("learn") = true);
 
+        py_HTM.def_property_readonly("connections", [](HTM_t& self)
```
Comment (breznak) [inline bindings/py/cpp_src/bindings/algorithms/py_HTM.cpp]: please do not expose TM.connections, it should have been protected: in c++ anyway. (I was trying to do that, but lots of tests hijack the functionality)

**Comment by breznak** [comment]
> Sooner or later, py_Connections is going to happen. 

ok, maybe, but the longer we have Connections "only for ourselves" in C++, the easier any changes are. 

So unless you have an urgent use-case, can we wait till someone requests and has a good need for py::Connections?

File: bindings/py/cpp_src/bindings/algorithms/py_HTM.cpp (hunk 3 above)
Comment (ctrl-z-9000-times) [reply bindings/py/cpp_src/bindings/algorithms/py_HTM.cpp]: Lots of tests here use that functionality too.
I think it's a legitimate use of the TM, to want to inspect the synapses as these tests do.

Comment (breznak) [reply bindings/py/cpp_src/bindings/algorithms/py_HTM.cpp]: ok, I now see this is read-only, so then yes, that's fine!

File: bindings/py/cpp_src/bindings/algorithms/py_Connections.cpp [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+ * See the GNU Affero Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses.
+ *
+ * http://numenta.org/licenses/
+ * --------------------------------------------------------------------- */
+
+/** @file
+ * PyBind11 bindings for Connections class
```
Comment (breznak) [inline bindings/py/cpp_src/bindings/algorithms/py_Connections.cpp]: this file already exists in the newer master, let's see if there's something new, or it merges cleanly.

File: src/nupic/algorithms/TemporalMemory.hpp [hunk 5]
```diff
@@ -512,6 +512,7 @@ using nupic::algorithms::connections::Synapse;
 
 public:
   Connections connections;
```
Comment (breznak) [inline src/nupic/algorithms/TemporalMemory.hpp]: can you please make public only a `const Connections&`, and use private `Connections connections_` ?

**Comment by ctrl-z-9000-times** [comment]
I think everything mentioned in the PR has now been done in a different PR.

**Pull request closed without merging**